File: packages/react-native-enriched-markdown/android/src/main/java/com/swmansion/enriched/markdown/utils/common/serialization/MarkdownASTSerializer.kt
package com.swmansion.enriched.markdown.utils.common.serialization

import com.swmansion.enriched.markdown.parser.MarkdownASTNode
import com.swmansion.enriched.markdown.parser.MarkdownASTNode.NodeType
import com.swmansion.enriched.markdown.utils.common.CodeBlockNode

object MarkdownASTSerializer {
  fun serializeNode(node: MarkdownASTNode): String {
    val buffer = StringBuilder()
    appendNode(node, buffer)
    return buffer.toString()
  }

  fun serializeChildren(node: MarkdownASTNode): String {
    val buffer = StringBuilder()
    for (child in node.children) {
      appendNode(child, buffer)
    }
    return buffer.toString()
  }

  /**
   * Serializes a Blockquote or Admonition node back to markdown, including
   * the `> ` prefix on every line and `> [!TYPE]` header for admonitions.
   */
  fun serializeBlockquote(node: MarkdownASTNode): String {
    val sb = StringBuilder()
    val isAdmonition = node.type == NodeType.Admonition
    if (isAdmonition) {
      val type = node.getAttribute("admonitionType")?.takeIf { it.isNotEmpty() } ?: "note"
      sb.append("> [!${type.uppercase(java.util.Locale.ROOT)}]\n")
    }
    val body = serializeBlockChildren(node)
    for (line in body.trimEnd('\n').split("\n")) {
      sb.append("> ")
      sb.append(line)
      sb.append("\n")
    }
    return sb.toString()
  }

  /**
   * Extracts the plain text content of an AST node recursively.
   * Inserts newlines between block-level children and handles soft/hard breaks.
   */
  fun plainText(node: MarkdownASTNode): String {
    val buffer = StringBuilder()
    appendPlainText(node, buffer)
    return buffer.toString().trim()
  }

  private fun appendPlainText(
    node: MarkdownASTNode,
    buffer: StringBuilder,
  ) {
    when (node.type) {
      NodeType.SoftBreak -> {
        buffer.append("\n")
      }

      NodeType.LineBreak -> {
        buffer.append("\n")
      }

      NodeType.BlankLine -> {
        val count = node.getAttribute("count")?.toIntOrNull() ?: 0
        repeat(count) { buffer.append("\n") }
      }

      NodeType.Paragraph -> {
        for (child in node.children) {
          appendPlainText(child, buffer)
        }
        buffer.append("\n")
      }

      NodeType.Heading -> {
        for (child in node.children) {
          appendPlainText(child, buffer)
        }
        buffer.append("\n")
      }

      else -> {
        buffer.append(node.content)
        for (child in node.children) {
          appendPlainText(child, buffer)
        }
      }
    }
  }

  // -- Block-level serialization --------------------------------------------------

  private fun serializeBlockChildren(node: MarkdownASTNode): String {
    val buffer = StringBuilder()
    val children = node.children
    for ((index, child) in children.withIndex()) {
      appendBlockNode(child, buffer)
      if (index < children.size - 1 && isBlockNode(child)) {
        buffer.append("\n")
      }
    }
    return buffer.toString()
  }

  private fun isBlockNode(node: MarkdownASTNode): Boolean =
    when (node.type) {
      NodeType.Paragraph,
      NodeType.Heading,
      NodeType.CodeBlock,
      NodeType.Blockquote,
      NodeType.Admonition,
      NodeType.Table,
      NodeType.UnorderedList,
      NodeType.OrderedList,
      NodeType.ThematicBreak,
      NodeType.BlankLine,
      NodeType.LatexMathDisplay,
      -> true

      else -> false
    }

  private fun appendBlockNode(
    node: MarkdownASTNode,
    buffer: StringBuilder,
  ) {
    when (node.type) {
      NodeType.Paragraph -> {
        appendChildren(node, buffer)
        buffer.append("\n")
      }

      NodeType.Heading -> {
        val level = (node.getAttribute("level") ?: "1").toIntOrNull() ?: 1
        buffer.append("#".repeat(level))
        buffer.append(" ")
        appendChildren(node, buffer)
        buffer.append("\n")
      }

      NodeType.CodeBlock -> {
        val code = CodeBlockNode.extractCode(node)
        val language = CodeBlockNode.language(node)
        val fenceChar = CodeBlockNode.fenceChar(node)
        buffer.append(CodeBlockNode.fencedMarkdown(code, language, fenceChar))
        buffer.append("\n")
      }

      NodeType.Blockquote, NodeType.Admonition -> {
        buffer.append(serializeBlockquote(node))
      }

      NodeType.ThematicBreak -> {
        buffer.append("---\n")
      }

      NodeType.BlankLine -> {
        val count = node.getAttribute("count")?.toIntOrNull() ?: 0
        repeat(count) { buffer.append("\n") }
      }

      NodeType.Table -> {
        appendTableMarkdown(node, buffer)
      }

      NodeType.UnorderedList -> {
        appendListMarkdown(node, buffer, ordered = false)
      }

      NodeType.OrderedList -> {
        appendListMarkdown(node, buffer, ordered = true)
      }

      NodeType.LatexMathDisplay -> {
        buffer.append("$$\n")
        appendChildren(node, buffer)
        buffer.append("\n$$\n")
      }

      else -> {
        appendChildren(node, buffer)
      }
    }
  }

  // Table markdown reconstruction from raw AST nodes. This parallels
  // TableContainerView.buildMarkdownFromRows() which works from processed
  // TableCellData; both must produce the same pipe-table format.
  private fun appendTableMarkdown(
    node: MarkdownASTNode,
    buffer: StringBuilder,
  ) {
    var headerDone = false
    for (section in node.children) {
      for (row in section.children) {
        if (row.type != NodeType.TableRow) continue
        val cells = row.children.map { serializeChildren(it) }
        buffer.append("| ")
        buffer.append(cells.joinToString(" | "))
        buffer.append(" |\n")
        if (!headerDone && row.children.any { it.type == NodeType.TableHeaderCell }) {
          val sep =
            row.children.map { cell ->
              when (cell.getAttribute("align")) {
                "center" -> ":---:"
                "right" -> "---:"
                else -> "---"
              }
            }
          buffer.append("| ")
          buffer.append(sep.joinToString(" | "))
          buffer.append(" |\n")
          headerDone = true
        }
      }
    }
  }

  private fun appendListMarkdown(
    node: MarkdownASTNode,
    buffer: StringBuilder,
    ordered: Boolean,
  ) {
    var number =
      if (ordered) {
        node.getAttribute("start")?.toIntOrNull()?.coerceAtLeast(0) ?: 1
      } else {
        1
      }
    for (item in node.children) {
      if (item.type != NodeType.ListItem) continue
      val prefix = if (ordered) "$number. " else "- "
      val content = serializeBlockChildren(item).trimEnd('\n')
      val lines = content.split("\n")
      lines.forEachIndexed { lineIdx, line ->
        if (lineIdx == 0) {
          buffer.append(prefix)
        } else {
          buffer.append(" ".repeat(prefix.length))
        }
        buffer.append(line)
        buffer.append("\n")
      }
      if (ordered) number++
    }
  }

  // -- Inline serialization -------------------------------------------------------

  private fun appendNode(
    node: MarkdownASTNode,
    buffer: StringBuilder,
  ) {
    when (node.type) {
      NodeType.Text -> {
        buffer.append(node.content)
      }

      NodeType.LineBreak -> {
        buffer.append("\\\n")
      }

      NodeType.SoftBreak -> {
        buffer.append("\n")
      }

      NodeType.Strong -> {
        buffer.append("**")
        appendChildren(node, buffer)
        buffer.append("**")
      }

      NodeType.Emphasis -> {
        buffer.append("*")
        appendChildren(node, buffer)
        buffer.append("*")
      }

      NodeType.Strikethrough -> {
        buffer.append("~~")
        appendChildren(node, buffer)
        buffer.append("~~")
      }

      NodeType.Underline -> {
        buffer.append("__")
        appendChildren(node, buffer)
        buffer.append("__")
      }

      NodeType.Superscript -> {
        buffer.append("^")
        appendChildren(node, buffer)
        buffer.append("^")
      }

      NodeType.Subscript -> {
        buffer.append("~")
        appendChildren(node, buffer)
        buffer.append("~")
      }

      NodeType.Highlight -> {
        buffer.append("==")
        appendChildren(node, buffer)
        buffer.append("==")
      }

      NodeType.Code -> {
        buffer.append("`")
        appendChildren(node, buffer)
        buffer.append("`")
      }

      NodeType.Link -> {
        val url = node.getAttribute("url") ?: ""
        buffer.append("[")
        appendChildren(node, buffer)
        buffer.append("](")
        buffer.append(url)
        buffer.append(")")
      }

      NodeType.Image -> {
        val alt = node.getAttribute("alt") ?: ""
        val url = node.getAttribute("url") ?: ""
        buffer.append("![")
        buffer.append(alt)
        buffer.append("](")
        buffer.append(url)
        buffer.append(")")
      }

      else -> {
        appendChildren(node, buffer)
      }
    }
  }

  private fun appendChildren(
    node: MarkdownASTNode,
    buffer: StringBuilder,
  ) {
    for (child in node.children) {
      appendNode(child, buffer)
    }
  }
}

File: packages/react-native-enriched-markdown/ios/utils/MarkdownASTSerializer.m
#import "MarkdownASTSerializer.h"
#import "ENRMCodeBlockContent.h"
#import "MarkdownASTNode.h"

static void serializeNode(MarkdownASTNode *node, NSMutableString *buffer);
static void serializeChildren(MarkdownASTNode *node, NSMutableString *buffer);
static void appendBlockNode(MarkdownASTNode *node, NSMutableString *buffer);
static void appendPlainText(MarkdownASTNode *node, NSMutableString *buffer);

static void serializeChildren(MarkdownASTNode *node, NSMutableString *buffer)
{
  for (MarkdownASTNode *child in node.children) {
    serializeNode(child, buffer);
  }
}

static void serializeNode(MarkdownASTNode *node, NSMutableString *buffer)
{
  if (!node)
    return;

  switch (node.type) {
    case MarkdownNodeTypeText:
      [buffer appendString:node.content ?: @""];
      break;

    case MarkdownNodeTypeLineBreak:
      [buffer appendString:@"\\\n"];
      break;

    case MarkdownNodeTypeSoftBreak:
      [buffer appendString:@"\n"];
      break;

    case MarkdownNodeTypeStrong:
      [buffer appendString:@"**"];
      serializeChildren(node, buffer);
      [buffer appendString:@"**"];
      break;

    case MarkdownNodeTypeEmphasis:
      [buffer appendString:@"*"];
      serializeChildren(node, buffer);
      [buffer appendString:@"*"];
      break;

    case MarkdownNodeTypeStrikethrough:
      [buffer appendString:@"~~"];
      serializeChildren(node, buffer);
      [buffer appendString:@"~~"];
      break;

    case MarkdownNodeTypeUnderline:
      [buffer appendString:@"__"];
      serializeChildren(node, buffer);
      [buffer appendString:@"__"];
      break;

    case MarkdownNodeTypeSuperscript:
      [buffer appendString:@"^"];
      serializeChildren(node, buffer);
      [buffer appendString:@"^"];
      break;

    case MarkdownNodeTypeSubscript:
      [buffer appendString:@"~"];
      serializeChildren(node, buffer);
      [buffer appendString:@"~"];
      break;

    case MarkdownNodeTypeHighlight:
      [buffer appendString:@"=="];
      serializeChildren(node, buffer);
      [buffer appendString:@"=="];
      break;

    case MarkdownNodeTypeCode:
      [buffer appendString:@"`"];
      serializeChildren(node, buffer);
      [buffer appendString:@"`"];
      break;

    case MarkdownNodeTypeLink: {
      NSString *url = node.attributes[@"url"] ?: @"";
      [buffer appendString:@"["];
      serializeChildren(node, buffer);
      [buffer appendFormat:@"](%@)", url];
      break;
    }

    case MarkdownNodeTypeImage: {
      NSString *alt = node.attributes[@"alt"] ?: @"";
      NSString *url = node.attributes[@"url"] ?: @"";
      [buffer appendFormat:@"![%@](%@)", alt, url];
      break;
    }

    case MarkdownNodeTypeParagraph:
    default:
      serializeChildren(node, buffer);
      break;
  }
}

// --- Block-level serialization ---

static BOOL isBlockNodeType(MarkdownNodeType type)
{
  switch (type) {
    case MarkdownNodeTypeParagraph:
    case MarkdownNodeTypeHeading:
    case MarkdownNodeTypeCodeBlock:
    case MarkdownNodeTypeBlockquote:
    case MarkdownNodeTypeAdmonition:
    case MarkdownNodeTypeTable:
    case MarkdownNodeTypeUnorderedList:
    case MarkdownNodeTypeOrderedList:
    case MarkdownNodeTypeThematicBreak:
    case MarkdownNodeTypeBlankLine:
    case MarkdownNodeTypeLatexMathDisplay:
      return YES;
    default:
      return NO;
  }
}

static void serializeBlockChildren(MarkdownASTNode *node, NSMutableString *buffer)
{
  NSArray<MarkdownASTNode *> *children = node.children;
  for (NSUInteger i = 0; i < children.count; i++) {
    MarkdownASTNode *child = children[i];
    appendBlockNode(child, buffer);
    if (i < children.count - 1 && isBlockNodeType(child.type)) {
      [buffer appendString:@"\n"];
    }
  }
}

// Table markdown reconstruction from raw AST nodes. This parallels
// TableContainerView's buildMarkdownFromRows which works from processed
// TableCellData; both must produce the same pipe-table format.
static void appendTableMarkdown(MarkdownASTNode *node, NSMutableString *buffer)
{
  BOOL headerDone = NO;
  for (MarkdownASTNode *section in node.children) {
    for (MarkdownASTNode *row in section.children) {
      if (row.type != MarkdownNodeTypeTableRow)
        continue;
      NSMutableArray<NSString *> *cells = [NSMutableArray arrayWithCapacity:row.children.count];
      for (MarkdownASTNode *cell in row.children) {
        [cells addObject:markdownFromASTNodeChildren(cell)];
      }
      [buffer appendFormat:@"| %@ |\n", [cells componentsJoinedByString:@" | "]];
      if (!headerDone) {
        BOOL isHeader = NO;
        for (MarkdownASTNode *cell in row.children) {
          if (cell.type == MarkdownNodeTypeTableHeaderCell) {
            isHeader = YES;
            break;
          }
        }
        if (isHeader) {
          NSMutableArray<NSString *> *sep = [NSMutableArray arrayWithCapacity:row.children.count];
          for (MarkdownASTNode *cell in row.children) {
            NSString *align = cell.attributes[@"align"];
            if ([align isEqualToString:@"center"])
              [sep addObject:@":---:"];
            else if ([align isEqualToString:@"right"])
              [sep addObject:@"---:"];
            else
              [sep addObject:@"---"];
          }
          [buffer appendFormat:@"| %@ |\n", [sep componentsJoinedByString:@" | "]];
          headerDone = YES;
        }
      }
    }
  }
}

static void appendListMarkdown(MarkdownASTNode *node, NSMutableString *buffer, BOOL ordered)
{
  NSUInteger number = 1;
  if (ordered) {
    NSString *startAttr = node.attributes[@"start"];
    if (startAttr != nil) {
      NSInteger parsed = startAttr.integerValue;
      number = (NSUInteger)MAX(0, parsed);
    }
  }
  for (MarkdownASTNode *item in node.children) {
    if (item.type != MarkdownNodeTypeListItem)
      continue;
    NSString *prefix = ordered ? [NSString stringWithFormat:@"%lu. ", (unsigned long)number] : @"- ";
    NSMutableString *content = [NSMutableString string];
    serializeBlockChildren(item, content);
    NSString *trimmed =
        content.length > 0 ? [content stringByTrimmingCharactersInSet:[NSCharacterSet newlineCharacterSet]] : @"";
    NSArray<NSString *> *lines = [trimmed componentsSeparatedByString:@"\n"];
    NSString *indent = [@"" stringByPaddingToLength:prefix.length withString:@" " startingAtIndex:0];
    for (NSUInteger l = 0; l < lines.count; l++) {
      [buffer appendString:(l == 0) ? prefix : indent];
      [buffer appendString:lines[l]];
      [buffer appendString:@"\n"];
    }
    if (ordered)
      number++;
  }
}

static void appendBlockNode(MarkdownASTNode *node, NSMutableString *buffer)
{
  if (!node)
    return;

  switch (node.type) {
    case MarkdownNodeTypeParagraph:
      serializeChildren(node, buffer);
      [buffer appendString:@"\n"];
      break;

    case MarkdownNodeTypeHeading: {
      NSInteger level = [node.attributes[@"level"] integerValue];
      if (level < 1)
        level = 1;
      for (NSInteger i = 0; i < level; i++)
        [buffer appendString:@"#"];
      [buffer appendString:@" "];
      serializeChildren(node, buffer);
      [buffer appendString:@"\n"];
      break;
    }

    case MarkdownNodeTypeCodeBlock: {
      NSString *code = ENRMCodeBlockExtractCode(node);
      NSString *language = ENRMCodeBlockLanguage(node);
      NSString *fenceChar = ENRMCodeBlockFenceChar(node);
      [buffer appendString:ENRMCodeBlockFencedMarkdown(code, language, fenceChar)];
      [buffer appendString:@"\n"];
      break;
    }

    case MarkdownNodeTypeBlockquote:
    case MarkdownNodeTypeAdmonition:
      [buffer appendString:markdownFromBlockquoteNode(node)];
      break;

    case MarkdownNodeTypeThematicBreak:
      [buffer appendString:@"---\n"];
      break;

    case MarkdownNodeTypeBlankLine: {
      NSInteger count = [node.attributes[@"count"] integerValue];
      for (NSInteger i = 0; i < count; i++) {
        [buffer appendString:@"\n"];
      }
      break;
    }

    case MarkdownNodeTypeTable:
      appendTableMarkdown(node, buffer);
      break;

    case MarkdownNodeTypeUnorderedList:
      appendListMarkdown(node, buffer, NO);
      break;

    case MarkdownNodeTypeOrderedList:
      appendListMarkdown(node, buffer, YES);
      break;

    case MarkdownNodeTypeLatexMathDisplay:
      [buffer appendString:@"$$\n"];
      serializeChildren(node, buffer);
      [buffer appendString:@"\n$$\n"];
      break;

    default:
      serializeChildren(node, buffer);
      break;
  }
}

static void appendPlainText(MarkdownASTNode *node, NSMutableString *buffer)
{
  switch (node.type) {
    case MarkdownNodeTypeSoftBreak:
    case MarkdownNodeTypeLineBreak:
      [buffer appendString:@"\n"];
      break;
    case MarkdownNodeTypeBlankLine: {
      NSInteger count = [node.attributes[@"count"] integerValue];
      for (NSInteger i = 0; i < count; i++) {
        [buffer appendString:@"\n"];
      }
      break;
    }
    case MarkdownNodeTypeParagraph:
    case MarkdownNodeTypeHeading:
      for (MarkdownASTNode *child in node.children) {
        appendPlainText(child, buffer);
      }
      [buffer appendString:@"\n"];
      break;
    default:
      if (node.content.length > 0) {
        [buffer appendString:node.content];
      }
      for (MarkdownASTNode *child in node.children) {
        appendPlainText(child, buffer);
      }
      break;
  }
}

// --- Public API ---

NSString *markdownFromASTNode(MarkdownASTNode *node)
{
  if (!node)
    return @"";
  NSMutableString *buffer = [NSMutableString string];
  serializeNode(node, buffer);
  return [buffer copy];
}

NSString *markdownFromASTNodeChildren(MarkdownASTNode *node)
{
  if (!node)
    return @"";
  NSMutableString *buffer = [NSMutableString string];
  serializeChildren(node, buffer);
  return [buffer copy];
}

NSString *markdownFromBlockquoteNode(MarkdownASTNode *node)
{
  if (!node)
    return @"";

  NSMutableString *result = [NSMutableString string];
  BOOL isAdmonition = (node.type == MarkdownNodeTypeAdmonition);

  if (isAdmonition) {
    NSString *type = node.attributes[@"admonitionType"];
    if (!type.length)
      type = @"note";
    [result appendFormat:@"> [!%@]\n", [type uppercaseString]];
  }

  NSMutableString *body = [NSMutableString string];
  serializeBlockChildren(node, body);

  NSString *trimmed = [body stringByTrimmingCharactersInSet:[NSCharacterSet newlineCharacterSet]];
  NSArray<NSString *> *lines = [trimmed componentsSeparatedByString:@"\n"];
  for (NSString *line in lines) {
    [result appendString:@"> "];
    [result appendString:line];
    [result appendString:@"\n"];
  }
  return [result copy];
}

NSString *plainTextFromASTNode(MarkdownASTNode *node)
{
  if (!node)
    return @"";
  NSMutableString *buffer = [NSMutableString string];
  appendPlainText(node, buffer);
  return [[buffer copy] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
}
